librz/magic/magic.c

```diff
@@ -237,100 +45,83 @@ RZ_API RzMagic *rz_magic_new(int flags) {
 		return NULL;
 	}
 	rz_magic_setflags(ms, flags);
-	ms->o.buf = ms->o.pbuf = NULL;
-	ms->c.li = malloc((ms->c.len = 10) * sizeof(*ms->c.li));
-	if (!ms->c.li) {
-		free(ms);
-		return NULL;
-	}
-	file_reset(ms);
-	ms->mlist = NULL;
-	ms->file = "unknown";
-	ms->line = 0;
 	return ms;
 }
 
-RZ_API void rz_magic_free(RzMagic *ms) {
-	if (ms) {
-		free_mlist(ms->mlist);
-		free(ms->o.pbuf);
-		free(ms->o.buf);
-		free(ms->c.li);
-		free(ms);
+RZ_API void rz_magic_free(RZ_NULLABLE RZ_OWN RzMagic *m) {
+	if (m) {
+		free(m->path);
+		rz_rbtree_free(m->magic_tree, magic_node_free_rb, NULL);
+		rz_rbtree_free(m->magic_named_tree, magic_node_free_rb, NULL);
+		free(m);
 	}
 }
 
-RZ_API bool rz_magic_load_buffer(RzMagic *ms, const char *magicdata) {
-	if (*magicdata == '#') {
-		struct mlist *ml = file_apprentice(ms, magicdata, FILE_LOAD);
-		if (ml) {
-			free_mlist(ms->mlist);
-			ms->mlist = ml;
-			return true;
-		}
-	} else {
-		eprintf("Magic buffers should start with #\n");
-	}
-	return false;
-}
+RZ_API bool rz_magic_load(RZ_NONNULL RZ_BORROW RzMagic *m, const char *magic_path) {
+	rz_return_val_if_fail(m, false);
 
-RZ_API bool rz_magic_load(RzMagic *ms, const char *magicfile) {
-	struct mlist *ml = file_apprentice(ms, magicfile, FILE_LOAD);
-	if (ml) {
-		free_mlist(ms->mlist);
-		ms->mlist = ml;
-		return true;
+	if (m->path) {
+		free(m->path);
 	}
-	return false;
-}
-
-RZ_API bool rz_magic_compile(RzMagic *ms, const char *magicfile) {
-	struct mlist *ml = file_apprentice(ms, magicfile, FILE_COMPILE);
-	free_mlist(ml);
-	return ml != NULL;
-}
-
-RZ_API bool rz_magic_check(RzMagic *ms, const char *magicfile) {
-	struct mlist *ml = file_apprentice(ms, magicfile, FILE_CHECK);
-	free_mlist(ml);
-	return ml != NULL;
-}
+	m->path = rz_str_dup(magic_path);
 
-RZ_API const char *rz_magic_descriptor(RzMagic *ms, int fd) {
-	return file_or_fd(ms, NULL, fd);
+	if (rz_file_is_regular(magic_path)) {
+		return magic_load_file(m, magic_path);
+	} else if (rz_file_is_directory(magic_path)) {
+		RzList *files = rz_sys_dir(magic_path);
+		if (!files) {
+			return false;
+		}
+		RzListIter *it;
+		const char *subname;
+		char *filepath = NULL;
+		RzStrBuf subpath;
+		rz_strbuf_init(&subpath);
+		bool result = true;
+		rz_list_foreach (files, it, subname) {
+			if (*subname == '.') {
+				continue;
+			}
+			if (!rz_str_cmp(subname, "..", -1)) {
+				continue;
+			}
+			filepath = rz_file_path_join(magic_path, subname);
+			result &= magic_load_file(m, filepath);
```

Comment:
Here I would add a warning if a specific file doesn't load correctly.
Otherwise the user won;t know which one in a directory failed.